File: services/insights/src/services/modelClient.ts
import { env } from "../env";

export type GeminiMessage = {
  role: "user" | "model";
  parts: Array<{ text: string }>;
};

export type ModelChatRequest = {
  systemInstruction: string;
  messages: GeminiMessage[];
};

export type ModelChatResponse = {
  text: string;
};

const GEMINI_URL = `https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash-lite:generateContent?key=${env.GEMINI_API_KEY}`;

export const modelClient = {
  async chat(input: ModelChatRequest): Promise<ModelChatResponse> {
    const body = {
      system_instruction: {
        parts: [{ text: input.systemInstruction }],
      },
      contents: input.messages,
    };

    const res = await fetch(GEMINI_URL, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });

    if (!res.ok) {
      const errorText = await res.text();
      throw new Error(`Gemini API error ${res.status}: ${errorText}`);
    }

    type GeminiResponse = {
      candidates?: Array<{
        content?: {
          parts?: Array<{ text?: string }>;
        };
      }>;
    };

    const data = (await res.json()) as GeminiResponse;
    const text = data.candidates?.[0]?.content?.parts?.[0]?.text ?? "";
    return { text };
  },
};

File: services/insights/rag_pipeline/model_client.py
"""Gemini API client for the RAG pipeline."""

from __future__ import annotations

from dataclasses import dataclass

import httpx

from .config import settings

_GEMINI_URL = (
    f"https://generativelanguage.googleapis.com/v1beta/models/"
    f"{settings.MODEL_NAME}:generateContent?key={settings.GEMINI_API_KEY}"
)

_client: httpx.AsyncClient | None = None


def _get_client() -> httpx.AsyncClient:
    global _client
    if _client is None:
        _client = httpx.AsyncClient(timeout=60.0)
    return _client


async def close_client() -> None:
    global _client
    if _client is not None:
        await _client.aclose()
        _client = None


@dataclass
class ModelResponse:
    text: str
    prompt_tokens: int | None = None
    response_tokens: int | None = None


async def chat(
    system_instruction: str,
    messages: list[dict],
) -> ModelResponse:
    """Call Gemini with system instruction + conversation history.

    `messages` should be a list of {"role": "user"|"model", "parts": [{"text": "..."}]}.
    """
    body = {
        "system_instruction": {"parts": [{"text": system_instruction}]},
        "contents": messages,
    }

    resp = await _get_client().post(
        _GEMINI_URL,
        json=body,
        headers={"Content-Type": "application/json"},
    )

    if not resp.is_success:
        raise RuntimeError(f"Gemini API error {resp.status_code}: {resp.text}")

    data = resp.json()
    text = (
        data.get("candidates", [{}])[0]
        .get("content", {})
        .get("parts", [{}])[0]
        .get("text", "")
    )

    # Extract token counts if available
    usage = data.get("usageMetadata", {})
    prompt_tokens = usage.get("promptTokenCount")
    response_tokens = usage.get("candidatesTokenCount")

    return ModelResponse(
        text=text,
        prompt_tokens=prompt_tokens,
        response_tokens=response_tokens,
    )
